docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
+  * nodejs 14.17.5: 0m5.603s
+  * rust 1.54.0: 0m13.587s
+  * golang 1.17: 1m30.428s
+
+* ``build-large``
+
+  * python 3.9.6: 2m33.688s
+  * mambaforge 4.10.1: 0m28.781s
+  * miniconda3 4.7.12: 0m10.551s
+  * nodejs 14.17.5: 0m6.136s
+  * rust 1.54.0: 0m14.716s
+  * golang 1.17: 1m36.470s
+
+Note that the only one that required compilation was Python.
+All the others, spent 100% of its time downloading the binary.
+These download times are *way better from EU* with a home internet connection.
+
+In the worst scenario: "none of the specified language version has a pre-built image",
+the build will require ~5 minutes to install all the language requirements.
+By providing *only* pre-built images with the Python version (that's the most time consuming),
+build times will only require ~2 minutes to install the others.
+However, requiring one version of each language is not a common case.
+
+Cache language binaries on S3
+-----------------------------
+
+``asdf`` scripts can be altered to download the ``.tar.gz`` dist files from a different mirror than the official one.
+Read the Docs can make usage of this to create a mirror hosted locally on S3 to get faster download speeds.
+This will make a good improvement for languages that offer binaries: ``nodejs``, ``rust`` and ``go``:
+
+* ``nodejs`` uses NODEJS_ORG_MIRROR: https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh#L5
+* ``rust`` uses RUSTUP_UPDATE_ROOT: https://github.com/rust-lang/rustup/blob/master/rustup-init.sh#L23
+* ``go`` has the URL hardcoded: https://github.com/kennyp/asdf-golang/blob/master/bin/download#L54
+
+However, currently Python does not offer binaries and a different solution is needed.
+Python versions can be pre-compiled once and expose the output on the S3 for the builders to download and extract in the correct PATH.
+
+
+Questions
+---------
+
+What Python versions will be pre-compiled and cached?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At start only a small subset of Python version will be pre-compiled:
+
+* 2.7.x
+* 3.7.x
+* 3.8.x
+* 3.9.x
+* 3.10.x
+* pypy3.x
+
+How do we upgrade a Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-How do we add/upgrade a Python version?
+Python patch versions can be upgraded by re-compiling the new patch version and making it available in our cache.
+For example, if 3.9.6 is the one available and 3.9.7 is released:
+
+* users specifying ``build.languages.python: "3.9"`` will get the newest version
+* users specifying ``build.languages.python: "3"`` will get the newest version
+
+.. note::
+
+   Python versions may need to be re-compiled each time that the ``-base`` image is re-built.
+   This is due that some underlying libraries that Python depend on may have changed.
+
+
+How long a Python pre-compiled version will be supported?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It will be supported as long as the ``-base`` Ubuntu OS version is supported by Read the Docs as well.
```

Comment:
I don't understand the answer to this question. Is this support for our usage, or for users?

Reply:
It refers to the support of a particular pre-compiled version that users can select in the config file.

As pre-compiled Python versions are compiled in a specific Ubuntu version, if we remove `ubuntu20-base` Docker images, the pre-compiled version may not work on `ubuntu22-base` because the base libraries have changed.

Reply:
Gotcha, but this is an implementation detail, noted for our benefit?

Reply:
Yeah, I will remove this question. It's not too relevant here.

Once we remove `ubuntu20-base` all the Python versions pre-compiled will go again together.